### PR TITLE
Use correct overload for Write-Object in Get-Variable

### PR DESCRIPTION
**Pull request by vexx32**
## PR Summary

Currently Get-Variable sends output to the pipeline without enumerating it when using `-ValueOnly`. This is at odds with almost every other cmdlet, and furthermore is at odds with the intuitive behaviour &mdash; I think it is reasonable to expect (indeed, that the _intent_ for this switch is for) the variable value to be returned in the same manner as it is when simply calling the variable.

In other words, it appears reasonable to expect that the two statements following the declaration should be equivalent:

```powershell
$a = 1, 2, 3, 4, 5
$a | Get-Member
Get-Variable -Name a -ValueOnly | Get-Member
```

If this behaviour is instead actually _intentional_, then it needs to be documented properly ([it isn't](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/get-variable?view=powershell-6#optional-parameters)), but I suspect it is more a matter of oversight than design here.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by iSazonov** [comment]
Formally it is a breaking change. /cc @SteveL-MSFT 

@mklement0 Could you please look the change too?

**Comment by iSazonov** [comment]
We don't cover the code by tests. 😕 
@vexx32 please add tests.

**Comment by mklement0** [comment]
The motivation for the change makes perfect sense to me and the change looks good. Nice catch, @vexx32.

**Comment by vexx32** [comment]
I can't really take the credit. @HumanEquivalentUnit stumbled across it trying to creatively troll someone on Reddit, apparently. 😂

I'll get some tests together during lunch. 😄

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs [hunk 1]
```diff
@@ -339,7 +339,7 @@ protected override void ProcessRecord()
                     matchFound = true;
                     if (_valueOnly)
                     {
-                        WriteObject(matchingVariable.Value);
+                        WriteObject(matchingVariable.Value, true);
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs]: I think this is more clear:
`WriteObject(matchingVariable.Value, enumerateCollection: true);`

**Review by PaulHigin — CHANGES_REQUESTED** [review]
Looks good if breaking change is approved and tests are added.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs (hunk 1 above)
Comment (vexx32) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs]: Yeah, I'll go ahead and do that instead. Thanks!

**Comment by vexx32** [comment]
@iSazonov @PaulHigin For a minute I was thinking "typical `Should` tests won't account for this... will they?"

And yet, despite a _horrifically_ confusing error message, they **do**:
image

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1 [hunk 2]
```diff
@@ -90,7 +90,15 @@ Describe "Get-Variable" -Tags "CI" {
 		Get-Variable -Name var1 -ValueOnly | Should -Be 4
 
 		Remove-Variable var1
-    }
+	}
+
+	It 'Should return the same value with -ValueOnly as directly calling the variable returns' {
```
Comment (PaulHigin) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1]: This test will be very confusing for someone if it fails.  I think that this should do an explicit check to ensure a Get-Variable -ValueOnly collection is enumerated to the pipe, and not depend on Pester 'Should'.  Maybe something similar to the original repro where the two cases are piped to Get-Member, then check type name.

Also, I think the 'It' description should be made more clear, or at least add a comment providing a clearer explanation of the test.  Something about -ValueOnly should enumerate collections to output pipe.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1 (hunk 2 above)
Comment (vexx32) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1]: Get-Member is a bit of a chunky way to go, generating a lot of data we don't need... we ought to be able to do `$v | % GetType` and `gv v | % GetType` and compare instead with the same result.

I'll try to clarify the `It` name. 😄

Comment (vexx32) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1]: Okay, I _think_ that ought to be clear enough... let me know if I need to rethink it. 😄

**Comment by iSazonov** [comment]
>And yet, despite a horrifically confusing error message, they do

Perhaps it could be reported to Pester.

**Comment by vexx32** [comment]
I have notified one of the maintainers both here and in the PowerShell Slack. I'll file an issue as well, just to be absolutely sure it's filed somewhere and not forgotten. 😄

**Comment by iSazonov** [comment]
Current test with checking types isn't good for me.

> Get-Variable -Name a -ValueOnly | Get-Member

I think a key in the test is that `Get-Member` has parameter with `ValueFromPipeline = true`
Seems we could use Compare-Object with DifferenceObject parameter like 
`Get-Variable -Name a -ValueOnly | Compare-Object -ReferenceObject $expected | Should ...`

**Comment by vexx32** [comment]
@iSazonov This is the only especially sensible method I can think of apart from the Get-Member test. Interestingly, `Compare-Object` does _not_ notice the difference; running that Compare-Object line (omitting the `Should`) gives no output. Appending `-IncludeEqual` indicates that it thinks both collections are equal and contain the same items.

Which is perfectly correct, I suppose, but it doesn't seem to distinguish the enumerated/not enumerated like one can by checking the type directly; indeed, it seems to me that this may be the only straightforward method to do so.

It's a bit of a tricky problem, and I acknowledge that there aren't a lot of good and easy ways to test it due to many cmdlets simply having an implicit *assumption* that piped objects will be enumerated and quietly handling those that don't, in an effort to make things like `$a | Compare-Object -ReferenceObject $b` and `Compare-Object -ReferenceObject $a -DifferenceObject $b` functionally equivalent.

**Comment by mklement0** [comment]
A pragmatic solution would be:

```powershell
(Get-Variable var1 -ValueOnly | Measure-Object).Count | Should -Be 3
```

If the array contained in `$var1` were unexpectedly sent as a single object, the result would be `1`.

**Comment by nohwnd** [comment]
I also think that the approach that @mklement0 proposed is the easiest and most reliable. The reason for the assertion failure is the same. Internally `Should -Be` compares [the count](https://github.com/pester/Pester/blob/master/Functions/Assertions/Be.ps1#L209) of items on both sides and fails when they differ. 

Depending on your version of Pester you might consider adding a `-Because` that would better explain what is the reason for the failure is.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1 [hunk 3]
```diff
@@ -90,7 +90,17 @@ Describe "Get-Variable" -Tags "CI" {
...
 		Remove-Variable var1
-    }
+	}
+
+	It 'Should enumerate its output in the same way that directly calling the value does' {
+		New-Variable -Name var1 -Value @(1, 2, 3)
+
+		# If output is not enumerated, Measure-Object will only see 1 object, not 3
+		$var1 | Measure-Object | Select-Object -ExpandProperty Count | Should -Be 3
+		Get-Variable -Name var1 -ValueOnly | Mesure-Object | Select-Object -ExpandProperty Count | Should -Be 3
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1]: We could remove line 99 and do
```powershell
Get-Variable -Name var1 -ValueOnly | Mesure-Object | Select-Object -ExpandProperty Count | Should -Be $var1.Count
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1 [hunk 4]
```diff
@@ -90,7 +90,19 @@ Describe "Get-Variable" -Tags "CI" {
...
+	It 'Should enumerate its output in the same way that directly calling the value does' {
+		New-Variable -Name var1 -Value @(1, 2, 3)
+
+		# If output is not enumerated, Measure-Object will only see 1 object, not 3
+        Get-Variable -Name var1 -ValueOnly |
+            Measure-Object |
+            Select-Object -ExpandProperty Count |
+			Should -Be $var1.Count
+
+		Remove-Variable -Name var1
```
[the remark this answers is not in the document]
Comment (vexx32) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1]: Huh, weird. I'm _sure_ I have VS Code to do spaces, but it was putting tabs in here, it seems. Not sure why, though. 😕

[the remark this answers is not in the document]
Comment (vexx32) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1]: Oh, wow, this file has tabs everywhere. Let me just... quickly fix that weirdness.

[the remark this answers is not in the document]
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1]: VS Code config from the repo set using spaces.

**Review by iSazonov — APPROVED** [review]

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by vexx32** [comment]
Pending committee decision.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this.  Agree that there is inconsistency, but there isn't clear value to taking this breaking change (likely breaking Pester usage which is highly impactful and may represent wider usage of existing behavior).

**Comment by SteveL-MSFT** [comment]
@vexx32 looking at your example, we believed that Pester relies on current behavior.  Correct us if that is not true.

**Comment by vexx32** [comment]
As far as I know (/cc @nohwnd as I'm sure he knows more than I where Pester's internals are concerned), Pester doesn't currently rely on `Get-Variable -ValueOnly` to compare items in any way.

Which example are you referring to? The example in the PR description doesn't have anything to do with Pester itself:

```powershell
$a = 1, 2, 3, 4, 5
$a | Get-Member
Get-Variable -Name a -ValueOnly | Get-Member
```

These simply differ because with Get-Variable (and as far as I know, apart from `ConvertFrom-Json` where it is intentional due to JSON structure, it's the **only** cmdlet that does this) doesn't enumerate its output, so there's no difference between `gv -val a | gm` and `gm -inp (gv -val a)`

The initial Pester test that I used to verify the behaviour is this:
```powershell
Get-Variable -Name var1 -ValueOnly | Should -Be $var1
```

And the reason this currently _fails_ is simply because `Should` is _expecting_ that a pipeline enumerates the output, so it **deliberately assumes** that any input that comes in implicitly has an already-discarded collection wrapper and matches things to the `-Be` provided parameter values accordingly.

The only way this breaks Pester itself would be if someone was using that exact test format and relying on the behaviour of `Get-Variable -ValueOnly` itself over the pipeline, which _may_ certainly be a concern, but to my knowledge Pester doesn't use that cmdlet specifically in this way, so it should remain unaffected. (Were I to need this, after all, I would opt instead for `Get-Variable` _without_ `-ValueOnly`, so that I could simply get the returned object and examine its `.Value` property, which is guaranteed to be true to life, pipeline or not.

If it did rely on that behaviour, I think I would expect the test to continue to fail once the change is made in the Get-Variable cmdlet, and we definitely saw the behaviour change as expected judging by the CI results and my own local tests.

I will of course respect the committee's decision, but I just want to make sure we're not retaining inconsistent behaviour without giving it a thorough look-see. 😄

**Comment by SteveL-MSFT** [comment]
@vexx32 thanks for the details.  I think the overall conclusion from @PowerShell/powershell-committee is that we don't know how much impact there is with users relying on the current behavior.  A simple search on GItHub turned up a number of scripts using `Get-Variable -ValueOnly`, but it's not easy to see how many use this in a pipeline (directly or indirectly).  The Pester remark may have been incorrect, but wasn't the deciding factor.  It's also not clear how much value this provides (other than consistency) to justify the breaking change.

**Comment by vexx32** [comment]
Roger that, I appreciate you taking the time to clarify! Personally, I think consistency is _super_ valuable but I also recognise that it's... dicey... at best to change long-standing behaviour for its sole sake. 😄

**Comment by nohwnd** [comment]
@vexx32 you analysis is correct. (If Pester relied on that behavior we could work around it by adding a case that would change based on the version of PowerShell, there are few conditional fixes and features like this already, but I also respect the decision of the committee of course :) )

**Comment by iSazonov** [comment]
Such fixes is good candidates for next LTS version if we'll produce LTS versions. We could mark such PRs with "Next-LTS" label.

**Pull request closed without merging**